File: In_class_Exercise_3/Library.py
class Book:
    def __init__(self, title, author, ISBN, price, stock):
        self.title = title
        self.author = author
        self.ISBN = ISBN
        self.price = price
        self.stock = stock

    def getTitle(self):
        return self.title

    def getAuthor(self):
        return self.author
    
    def borrow(self):
        if self.stock > 0:
            self.stock -= 1
            return True
        else:
            return False
        
class Library:
    def __init__(self):
        self.books = []
    
    def addBook(self, book):
        self.books.append(book)
    
    def searchBook(self, keyword):
        found = False
        for b in self.books:
            if b.getTitle() == keyword or b.getAuthor() == keyword:
                print(f"Found: {b.getTitle()} by {b.getAuthor()}")
                found = True
        if not found:
            print("No matching books found.")

    def borrowBook(self, title):
        for b in self.books:
            if b.getTitle() == title:
                if b.borrow():
                    print("Success! You borrowed")
                else:
                    print("Failed! Out of stock.")
                return
        print("Book not found in the library.")

# Main function
if __name__ == "__main__":
    library = Library()
    
    # Adding books
    book1 = Book("1984", "George Orwell", "1234567890", 9.99, 3)
    book2 = Book("To Kill a Mockingbird", "Harper Lee", "0987654321", 12.99, 2)
    library.addBook(book1)
    library.addBook(book2)
    
    # Searching for a book
    library.searchBook("1984")
    library.searchBook("J.K. Rowling")
    
    # Borrowing books
    library.borrowBook("1984")
    library.borrowBook("1984")
    library.borrowBook("1984")
    library.borrowBook("1984")  # This should indicate out of stock
    library.borrowBook("The Great Gatsby")  # This should indicate book not found

File: In_class_Exercise_3/Library.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

// 1. Class Book
class Book {
private:
    string title;
    string author;
    string ISBN;
    double price;
    int stock;

public:
    Book(string t, string a, string i, double p, int s) 
        : title(t), author(a), ISBN(i), price(p), stock(s) {}

    string getTitle() { return title; }
    string getAuthor() { return author; }
    
    bool borrow() {
        if (stock > 0) {
            stock--;
            return true;
        }
        return false;
    }

    void display() {
        cout << "Title: " << title << " | Author: " << author << " | Stock: " << stock << endl;
    }
};

// 2. Class Library
class Library {
private:
    vector<Book> books;

public:
    // a. Add new books
    void addBook(Book b) {
        books.push_back(b);
        cout << "-> Added: " + b.getTitle() << endl;
    }

    // b. Search
    void searchBook(string keyword) {
        bool found = false;
        for (auto &b : books) {
            if (b.getTitle() == keyword || b.getAuthor() == keyword) {
                b.display();
                found = true;
            }
        }
        if (!found) cout << "No book found." << endl;
    }

    // c. Borrow
    void borrowBook(string title) {
        for (auto &b : books) {
            if (b.getTitle() == title) {
                if (b.borrow()) {
                    cout << "Success! You borrowed" << endl;
                } else {
                    cout << "Failed! Out of stock." << endl;
                }
                return;
            }
        }
        cout << "Book not found." << endl;
    }
};

// 3. Main
int main() {
    Library myLib;

    myLib.addBook(Book("Harry Potter", "J.K. Rowling", "12345", 20.0, 3));
    myLib.addBook(Book("The Hobbit", "Tolkien", "67890", 15.0, 0)); // Hết hàng

    myLib.searchBook("Harry Potter");

    myLib.borrowBook("Harry Potter"); // OK
    myLib.borrowBook("The Hobbit");   // Fail (Stock = 0)

    return 0;
}